gapic/src/main/com/google/gapid/widgets/ScenePanel.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (C) 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.gapid.widgets;
+
+import com.google.gapid.glcanvas.GlCanvas;
+
+import com.google.gapid.glviewer.gl.Renderer;
+import com.google.gapid.glviewer.gl.Scene;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.internal.DPIUtil;
+import org.eclipse.swt.layout.FillLayout;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Listener;
+import org.lwjgl.opengl.GL;
+import org.lwjgl.opengl.GLCapabilities;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * {@link GlCanvas} used to render a {@link Scene}.
+ */
+public class ScenePanel<T> extends GlCanvas {
+  private final GLCapabilities caps;
+  private final Renderer renderer;
+  private final Scene scene;
+  private final AtomicReference<T> sceneData = new AtomicReference<>();
+  private Map<Integer, List<Listener>> eventListeners;
+  private int numSuspendedUpdates;
+  private boolean isPendingUpdate;
+  private boolean isPendingRender;
+
+  public ScenePanel(Composite parent, Scene<T> scene) {
+    super(parent, SWT.NO_BACKGROUND);
+    setLayout(new FillLayout(SWT.VERTICAL));
+    this.scene = scene;
+    renderer = new Renderer();
+
+    setCurrent();
+    caps = GL.createCapabilities();
+    initialize();
+
+    addListener(SWT.Resize, e -> {});
```

Comment:
Add a comment inside the empty braces that indicate that this here only because there's an implicit side effect. When I just read this, I was about to comment to remove this as it looks like it has no effect...